pycloudlib/qemu/cloud.py

```diff
@@ -0,0 +1,633 @@
+            instance_dir,
+            image_id,
+        )
+
+        # We make a COW image from the base image so we can make destructive
+        # changes without modifying the original image
+        instance_path = instance_dir / "inst.qcow2"
+        self._create_qcow_image(
+            instance_path=instance_path, base_image=base_image
+        )
+
+        ssh_port = get_free_port()
+        telnet_port = get_free_port()
+        socket_path = instance_dir / "qmp-socket"
+        qemu_args = [
+            self.qemu_binary,
+            "-enable-kvm",
+            "-cpu",
+            "host",
+            "-smp",
+            f"cpus={cpus}",
+            "-m",
+            f"size={memory}",
+            "-net",
+            "nic",
+            "-net",
+            f"user,hostfwd=tcp::{ssh_port}-:22",
+            "-hda",
+            str(instance_path),
+            "-nographic",
+            "-chardev",
+            f"socket,id=char0,logfile={instance_dir / 'console.log'},"
+            f"host=0.0.0.0,port={telnet_port},telnet=on,server=on,"
+            "wait=off,mux=on",
+            "-serial",
+            "chardev:char0",
+            "-qmp",
+            f"unix:{socket_path},server,wait=off",
+            "-no-shutdown",
+        ]
+
+        if not no_seed_iso:
+            # By default, even if no user data is passed, we'll need to
+            # use NoCloud and create a seed iso, otherwise we have no
+            # way of accessing our instance
+            seed_path = self._create_seed_iso(
+                instance_dir=instance_dir,
+                user_data=add_key_to_cloud_config(
+                    public_key=self.key_pair.public_key_content,
+                    user_data=user_data,
+                ),
+                meta_data=meta_data,
+                vendor_data=vendor_data,
+            )
+
+            if seed_path:
+                driver = f"driver=raw,file={str(seed_path)},if=virtio"
+                qemu_args.extend(["-drive", driver])
+
+        if kernel_cmdline and not kernel_path:
+            kernel_path = self._get_kernel_path(
+                kernel_path=kernel_path,
+                kernel_cmdline=kernel_cmdline,
+                image_id=image_id,
+                base_image=base_image,
+            )
+
+        if kernel_path:
+            kernel_cmdline = self._update_kernel_cmdline(
+                kernel_cmdline=kernel_cmdline
+            )
+            qemu_args.extend(
+                ["-kernel", str(kernel_path), "-append", kernel_cmdline]
+            )
+
+        if launch_args:
+            qemu_args.extend(launch_args)
+
+        self._log.info("Launching qemu with args: %s", qemu_args)
+        handle = subprocess.Popen(  # pylint: disable=R1732
+            qemu_args,
+            stdin=subprocess.DEVNULL,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL,
+        )
+        instance = QemuInstance(
```

Comment:
Isn't this true of all of our clouds? `delete()` has to be called (either manually or via context manager) or the resource is leaked. Do you think that it's a bigger problem here?

Reply:
I was thinking in terms of OS resources. `Popen` could leak file descriptors if `kill` / `close` is not called. So imagine we have a long-running process creating `QemuInstance`s but not calling `QemuInstance.delete`, if those instances get out of scope and garbage collected, then the qemu process created here is going to be leaked I think.

I was suggesting to clean up this process like:

```py
class QemuInstance:
    def __del__(self):
        self.handle.close()
```

Reply:
So, the difference with other clouds is that in this case the parent process could leak memory/fds and exhaust the host, but maybe I am overthinking here. What do you think?

Reply:
Yeah, I see your point. It'll definitely be a lot easier to overwhelm a host with VMs vs the cost of leaking cloud instances. I'd be most worried about RAM as using DEVNULL means we won't be allocating new fds every invocation.

On the other side of the argument though, I would like to be able to use the `KEEP_INSTANCE = True` functionality that we have in cloud-init to debug instances after the fact. In this case, it would make sense to let processes live even longer than they currently do (i.e., past the life of the pycloudlib process).

I think there are trade-offs both ways. I'll bring it up at standup.

Reply:
As discussed in standup, we'll go ahead with letting the process live until explicitly killed.